jupyterhub/tests/conftest.py

```diff
@@ -296,6 +311,10 @@ def _mockservice(request, app, url=False):
         else:
             spec['url'] = 'http://127.0.0.1:%i' % random_port()
 
+    if external:
+
+        spec['oauth_redirect_uri'] = 'http://127.0.0.1:%i' % random_port()
```

Comment:
I always get a bit confused about the `%s`, `%i`, `%d`, `%r` etc.

I understand that this statement with `%i` is equivalent to `%d`, and it seems to be acting like `f{random_port():.0f}`.

I see

```python
def random_port():
    """Get a single random port."""
    sock = socket.socket()
    sock.bind(('', 0))
    port = sock.getsockname()[1]
    sock.close()
    return port
```

I assume random_port is always a integer, but is this protection for if it is a float? I note that both f strings with `:.0f` and `%i` will error if a string or `None` is given.

Reply:
Is there relevant intent to use/not use an f-string like the one below where we don't coerce a float to an int, and is there relevant intent to use/not using an f-string with `.0f` formatting instead of `%i`?

```
        spec['oauth_redirect_uri'] = f'http://127.0.0.1:{random_port()}'
```

I'm asking to learn if I need to adopt certain patterns for various situations or if doesn't matter.

Reply:
I wouldn't call it protection, it's just the idiomatic way to put an integer in a string in the old percent-style (I've been writing it that way for a long time!). I might have used an f-string if I were typing from scratch, but you can see this is a copy/paste of the line just above. If I used an f-string, I wouldn't add a type formatter (If I did, it would be `f"{random_port():d}"` where 'd' is for [decimal integer](https://docs.python.org/3/library/string.html#format-specification-mini-language). But I'd trust the return type of the function in this case.

I wouldn't use a float format string (`.0f`) for something like a port that's an integer by definition.

Reply:
> I'm asking to learn if I need to adopt certain patterns for various situations or if doesn't matter.

It doesn't matter. It's just me copying pre-f-string code / habits. An f-string with no formatter would be AOK here, for me. I've just written `host:%i' % port` so many times in my life, it's hard to stop!